### PR TITLE
Require applications to be referenced by name instead of ID

### PR DESCRIPTION
**Pull request by pb-seiso**
# Contributor Comments

- Added get_app_id() utility function to allow users to pass application name instead of ID
- Updated configurations to accept app_name instead of app_id
- Added unit tests for get_app_id and updated existing tests to use it

## Pull Request Checklist
Thank you for submitting a contribution to `easy_sast`!

In order to streamline the review of your contribution we ask that you review and comply with the below requirements:
 - [x] Ensure that `make build` completes successfully
 - [x] Keep our reports up-to-date by committing the output of `make report`
 - [x] Rebase your branch against the latest commit of the target branch, which likely should be `master`
 - [x] If there is an issue associated with your Pull Request, align your PR title with [this documentation](https://help.github.com/en/articles/closing-issues-using-keywords)

### CODE REVIEW

**Comment by JonZeolla** [comment]
@pb-seiso merge conflict.  Let's hold on this until #36 is in, then rebase on master and take it from there.

**Comment by pb-seiso** [comment]
@SeisoLLC/cto Rebased on master, fixed conflicts, updated reports, and ran a quick end-to-end to make sure everything still looks good. Let me know if you have any other feedback.

**Comment by pb-seiso** [comment]
@JonZeolla Are you waiting for anything else from me before reviewing this?

**Comment by JonZeolla** [comment]
No, sorry it's on my backlog

**Comment by JonZeolla** [comment]
@pb-seiso please rebase on `master`

**Comment by pb-seiso** [comment]
> @pb-seiso please rebase on `master`

Done, ready for your review.

File: tests/test_api.py [hunk 1]
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
     ## VeracodeXMLAPI app_id property
     def test_veracode_xml_api_app_id(self):
         """
-        Test the VeracodeXMLAPI app_id property
+        Test the VeracodeXMLAPI app_name property
```
Comment (JonZeolla) [inline tests/test_api.py:66]:
```suggestion
        Test the VeracodeXMLAPI app_id property
```

File: tests/test_api.py [hunk 2]
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
...
         """
-        veracode_xml_api = VeracodeXMLAPI()
+        with patch("veracode.api.get_app_id", return_value=0):
+            veracode_xml_api = VeracodeXMLAPI("TestApp")
 
         # Fail when attempting to get the app_id property because app_id is
```
Comment (JonZeolla) [inline tests/test_api.py:71]: This is no longer true as of this PR.

File: tests/test_api.py [hunk 3]
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
     ## VeracodeXMLAPI app_id property
     def test_veracode_xml_api_app_id(self):
         """
-        Test the VeracodeXMLAPI app_id property
+        Test the VeracodeXMLAPI app_name property
         """
-        veracode_xml_api = VeracodeXMLAPI()
+        with patch("veracode.api.get_app_id", return_value=0):
```
Comment (JonZeolla) [inline tests/test_api.py:68]: `app_id` should be a string, not an int.

File: tests/constants.py [hunk 4]
```diff
@@ -769,6 +777,43 @@
...
+VALID_APPLIST_API["base_url"] = "https://analysiscenter.veracode.com/api/"
+VALID_APPLIST_API["version"] = {"getapplist.do": "5.0"}
+
+VALID_APPLIST_API["app_id"] = "1337"
+VALID_APPLIST_API["app_name"] = "TestApp"
+VALID_APPLIST_API["build_id"] = "v1.2.3"
+VALID_APPLIST_API["username"] = "TestUser"
+
+# https://help.veracode.com/reader/LMv_dtSHyb7iIxAQznC~9w/kb2SM9net26_L91VploQGw
+VALID_APPLIST_API_RESPONSE_XML = {}
```
Comment (JonZeolla) [inline tests/constants.py:792]: Replicate throughout.
```suggestion
VALID_UPLOAD_API_APPLIST_RESPONSE_XML = {}
```

File: tests/test_api.py [hunk 5]
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
...
+            veracode_xml_api = VeracodeXMLAPI("TestApp")
```
Comment (JonZeolla) [inline tests/test_api.py:69]: Where possible, be specific
```suggestion
            veracode_xml_api = VeracodeXMLAPI(app_name="TestApp")
```

File: tests/test_api.py [hunk 6]
```diff
@@ -62,15 +63,16 @@ def test_veracode_xml_api_version(self):
...
 
         # Fail when attempting to get the app_id property because app_id is
         # hard coded to an invalid value to discourage direct use of this class
         self.assertRaises(ValueError, getattr, veracode_xml_api, "app_id")
 
-        # Fail when attempting to set the app_id property to an invalid value
+        # Fail when attempting to set the app_name property to an invalid value
```
Comment (JonZeolla) [inline tests/test_api.py:75]:
```suggestion
        # Fail when attempting to set the app_id property to an invalid value
```

File: tests/test_api.py [hunk 7]
```diff
@@ -368,7 +421,7 @@ def test_upload_api_sandbox_id(self):
         self.assertIsNone(setattr(upload_api, "sandbox_id", None))
         self.assertIsNone(setattr(upload_api, "sandbox_id", "12489"))
 
-        # Succeed when getting a valid app_id property
+        # Succeed when getting a valid sandbox_id property
```
Comment (JonZeolla) [inline tests/test_api.py:424]: Nice catch

File: tests/test_config.py [hunk 8]
```diff
@@ -732,33 +732,48 @@ def test_apply_config(self):
...
         # Ensure calling the apply_config function with different `app_id`s
         # does not result in an object with the same version string (which is
         # unique per-API)
-        upload_api = UploadAPI(app_id="31337")
-        applied_upload_api = config.apply_config(api=upload_api, config=configuration)
-        # Note: This must be different than the above app_id
-        results_api = ResultsAPI(app_id="1337")
-        applied_results_api = config.apply_config(api=results_api, config=configuration)
-        self.assertNotEqual(applied_results_api.version, applied_upload_api.version)
+        with patch("veracode.api.get_app_id", return_value="1337"):
```
Comment (JonZeolla) [inline tests/test_config.py:763]: See the comment below `# Note: This must be different than the above app_id`
```suggestion
        with patch("veracode.api.get_app_id", return_value="31337"):
```

File: tests/test_api.py [hunk 9]
```diff
@@ -35,7 +35,8 @@ def test_veracode_xml_api_version(self):
         """
         Test the VeracodeXMLAPI version property
         """
-        veracode_xml_api = VeracodeXMLAPI()
+        with patch("veracode.api.get_app_id", return_value="1337"):
```
Comment (JonZeolla) [inline tests/test_api.py:38]: Please do this in the 90 bazillion places across the various tests.
```suggestion
        with patch("veracode.api.get_app_id", return_value=constants.VALID_UPLOAD_API["app_id"]):
```

File: veracode/api.py [hunk 10]
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
     https://help.veracode.com/reader/LMv_dtSHyb7iIxAQznC~9w/pd_p6JjB9PcDNH3GzWF5Ag
     """
 
-    def __init__(self):
-        # Hard code these to None as they should be specified in the derived classes
-        self._app_id = None
+    def __init__(self, app_name: str):
```
Comment (JonZeolla) [inline veracode/api.py:31]: Require kwarg
```suggestion
    def __init__(self, *, app_name: str):
```

File: veracode/api.py [hunk 11]
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
     https://help.veracode.com/reader/LMv_dtSHyb7iIxAQznC~9w/pd_p6JjB9PcDNH3GzWF5Ag
     """
 
-    def __init__(self):
-        # Hard code these to None as they should be specified in the derived classes
```
Comment (JonZeolla) [inline veracode/api.py:28]: This is still true of `self._version = None`, reword and keep the comment

File: veracode/api.py [hunk 12]
```diff
@@ -157,12 +188,11 @@ class UploadAPI(VeracodeXMLAPI):  # pylint: disable=too-many-instance-attributes
     A class to interact with the Upload API
     """
 
-    def __init__(self, *, app_id: str):
+    def __init__(self, *, app_name: str):
         # Don't forget to call the init of the parent
-        super().__init__()
+        super().__init__(app_name)
```
Comment (JonZeolla) [inline veracode/api.py:193]:
```suggestion
        super().__init__(app_name=app_name)
```

File: veracode/api.py [hunk 13]
```diff
@@ -336,12 +366,11 @@ class ResultsAPI(VeracodeXMLAPI):
     A class to interact with the Results API
     """
 
-    def __init__(self, *, app_id: str):
+    def __init__(self, *, app_name: str):
         # Don't forget to call the init of the parent
-        super().__init__()
+        super().__init__(app_name)
```
Comment (JonZeolla) [inline veracode/api.py:371]:
```suggestion
        super().__init__(app_name=app_name)
```

File: veracode/api.py [hunk 14]
```diff
@@ -389,12 +418,10 @@ class SandboxAPI(VeracodeXMLAPI):
     A class to interact with the Sandbox API
     """
 
-    def __init__(self, *, app_id: str, sandbox_name: str):
+    def __init__(self, *, app_name: str, sandbox_name: str):
         # Don't forget to call the init of the parent
-        super().__init__()
+        super().__init__(app_name)
```
Comment (JonZeolla) [inline veracode/api.py:423]:
```suggestion
        super().__init__(app_name=app_name)
```

File: veracode/api.py [hunk 15]
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
...
-        self._app_id = None
+    def __init__(self, app_name: str):
         self._version = None
 
         ## Use the setter to apply a default to ensure it is valid
         self.base_url = "https://analysiscenter.veracode.com/api/"
 
+        # Set app name and look up ID
+        self._app_name = app_name
+        self._app_id = get_app_id(app_name)
```
Comment (JonZeolla) [inline veracode/api.py:39]: This means we now globally require on the upload API `getapplist.do` endpoint.  Please document here, in sandbox api, results api, and in markdown.

File: veracode/api.py [hunk 16]
```diff
@@ -24,14 +28,16 @@ class VeracodeXMLAPI:
...
-        # Hard code these to None as they should be specified in the derived classes
-        self._app_id = None
+    def __init__(self, app_name: str):
         self._version = None
 
         ## Use the setter to apply a default to ensure it is valid
         self.base_url = "https://analysiscenter.veracode.com/api/"
 
+        # Set app name and look up ID
+        self._app_name = app_name
```
Comment (JonZeolla) [inline veracode/api.py:38]: Shouldn't we be using the setter to ensure it's valid?
```suggestion
        self.app_name = app_name
```

File: veracode/api.py (hunk 15 above)
Comment (JonZeolla) [inline veracode/api.py:39]: Using the setter to ensure it's valid.
```suggestion
        self.app_id = get_app_id(app_name)
```

File: veracode/utils.py [hunk 17]
```diff
@@ -393,3 +401,41 @@ def is_valid_netloc(*, netloc: str) -> bool:
...
+
+
+def get_app_id(app_name: str) -> Union[int, None]:
+    """
+    Query for and return the app_id associated with the app_name
+
+    https://help.veracode.com/reader/orRWez4I0tnZNaA_i0zn9g/Z4Ecf1fw7868vYPVgkglww
+    """
+    try:
+        ## Use the setter to apply a default to ensure it is valid
```
Comment (JonZeolla) [inline veracode/utils.py:413]: I feel like this is a misplaced comment.

File: veracode/utils.py [hunk 18]
```diff
@@ -393,3 +401,41 @@ def is_valid_netloc(*, netloc: str) -> bool:
         return True
 
     return False
+
+
+def get_app_id(app_name: str) -> Union[int, None]:
```
Comment (JonZeolla) [inline veracode/utils.py:406]:
```suggestion
@validate
def get_app_id(app_name: str) -> Union[int, None]:
```

File: veracode/utils.py [hunk 19]
```diff
@@ -393,3 +401,41 @@ def is_valid_netloc(*, netloc: str) -> bool:
...
+    """
+    Query for and return the app_id associated with the app_name
+
+    https://help.veracode.com/reader/orRWez4I0tnZNaA_i0zn9g/Z4Ecf1fw7868vYPVgkglww
+    """
+    try:
+        ## Use the setter to apply a default to ensure it is valid
+        version = {"getapplist.do": "5.0"}
```
Comment (JonZeolla) [inline veracode/utils.py:414]: I don't want two upload api version dicts; can we abstract this into `constants.py` for both this and `UploadAPI`'s `self.version`?  Perhaps for consistency we also move the `SandboxAPI` and `ResultsAPI` `self.version` `dict`s.

File: veracode/utils.py [hunk 20]
```diff
@@ -393,3 +401,41 @@ def is_valid_netloc(*, netloc: str) -> bool:
...
+        base_url = "https://analysiscenter.veracode.com/api/"
```
Comment (JonZeolla) [inline veracode/utils.py:415]: Please move to `constants.py` and use here and in `VeracodeXMLAPI`'s `__init__`

**Review by JonZeolla — CHANGES_REQUESTED** [review]
Nice work; some polish needed then should be good to go.  If you don't mind, please add my suggestions to a batch and commit them all together (if you agree with them) so I don't have two dozen new commits to sift through.

File: tests/test_api.py (hunk 3 above)
Comment (pb-seiso) [reply tests/test_api.py:68]: Looks like I originally made `get_app_id` return `Union[int, None]` - I'm assuming that needs updated to `Union[str, None]` then? 

The XML schema says the app_id is a `long` which caused some confusion I think: https://analysiscenter.veracode.com/resource/2.0/applist.xsd

Comment (JonZeolla) [reply tests/test_api.py:68]: Right, https://github.com/SeisoLLC/easy_sast/blob/e220a85c7cdd2594e2f43c11f595a6d3d739175e/tests/constants.py#L49 shows my explicit test to ensure this isn't an int.  This is an example of the variation that I noted at https://github.com/SeisoLLC/easy_sast/blob/e220a85c7cdd2594e2f43c11f595a6d3d739175e/tests/constants.py#L81

**Comment by pb-seiso** [comment]
@JonZeolla This is ready again for review. The rebase on `master` prevented a nice commit batch and I'll have to rebase again anyway to fix commit messages and reorder them each onto `HEAD`. You can just review [this one](https://github.com/SeisoLLC/easy_sast/pull/37/commits/8af94510b7e8c3e164ea0ebc91aecfb921afebbc).

Note there will be a lint error on #43 since the commit lacks a body message. Please either fix that upstream or let me know if you want me to edit it when I rebase.

File: tests/test_utils.py [hunk 21]
```diff
@@ -49,7 +49,13 @@ def test_results_function(
         ) -> None:
             pass
 
-        results_api = ResultsAPI(app_id=test_constants.VALID_RESULTS_API["app_id"])
+        with patch(
+            "veracode.api.get_app_id",
+            return_value=test_constants.VALID_UPLOAD_API["app_id"],
```
Comment (JonZeolla) [inline tests/test_utils.py:54]: It looks like we're mixing `VALID_UPLOAD_API` and `VALID_RESULTS_API` here and below, please make sure they're aligned
```suggestion
            return_value=test_constants.VALID_RESULTS_API["app_id"],
```

**Review by JonZeolla — CHANGES_REQUESTED** [review]
Just one minor item

File: veracode/api.py (hunk 10 above)
Comment (JonZeolla) [reply veracode/api.py]: This was set as resolved but I don't see it resolved

File: veracode/api.py (hunk 15 above)
Comment (JonZeolla) [reply veracode/api.py]: Please document this somewhere

**Comment by JonZeolla** [comment]
@pb-seiso this is good to go now; force pushed a commit rewording.  A simple conflict resolution is to just accept them as-is then re-generate them all.

File: veracode/api.py (hunk 15 above)
Comment (pb-seiso) [reply veracode/api.py]: Note to self, make a README / wiki edit:
`Require permissions to access UploadAPI along with the other APIs, as get_app_id requires it for app ID resolution.`

File: tests/test_utils.py (hunk 21 above)
Comment (pb-seiso) [reply tests/test_utils.py:54]: Should be fixed now.

**Comment by pb-seiso** [comment]
@JonZeolla This is rebased and ready again for review. I ran basic integration tests on the `app_name` parameter which looked good, but please feel free to re-test.

Note that I've also included a fix to change the Docker image to one that uses Python 3.9.

**Review by JonZeolla — APPROVED** [review]
Nice work!  Man, is it nice to see this thing get wrapped up.  I will give you the pleasure of the squash merge; make sure to leave a properly formatted commit message for git linting.

**Pull request merged**